### PR TITLE
[Staging] PLT-415: Implementation of task delete and create API 

### PR DESCRIPTION
**Pull request by sumandas0**
## Task Delete and Create API

> We need to clear multiple tasks in some customer instances but before that we need to implement few APIs to do those operations

## Type of change
- [x] New feature (adds functionality)

Linear: https://linear.app/atlanproduct/issue/PLT-415/clear-tasks-in-customer-instances-with-large-no-of-pending-tasks

### CODE REVIEW

File: repository/src/main/java/org/apache/atlas/tasks/AtlasTaskService.java [hunk 1]
```diff
@@ -124,6 +133,197 @@ public void retryTask(String taskGuid) throws AtlasBaseException {
...
+        if (classificationVertex != null) {
+            ret = classificationVertex.getIdForDisplay();
+        }
+
+        return ret;
+    }
+
+    @Override
+    public List<AtlasTask> deleteAtlasTasks(List<AtlasTask> tasks) {
+        AtlasPerfMetrics.MetricRecorder metric = RequestContext.get().startMetricRecord("deleteAtlasTasks");
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/tasks/AtlasTaskService.java:213]: Move metric recording to [TaskREST.java]

File: repository/src/main/java/org/apache/atlas/tasks/AtlasTaskService.java [hunk 2]
```diff
@@ -124,6 +133,197 @@ public void retryTask(String taskGuid) throws AtlasBaseException {
...
+
+            if (results.hasNext()) {
+                graph.removeVertex(results.next());
+            }
+        } catch (Exception exception) {
+            LOG.error("Error: deletingByGuid: {}", guid);
+
+            throw new AtlasBaseException(exception);
+        } finally {
+            graph.commit();
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/tasks/AtlasTaskService.java:299]: Do we need to do graph.commit() manually?

File: repository/src/main/java/org/apache/atlas/tasks/AtlasTaskService.java [hunk 3]
```diff
@@ -124,6 +133,197 @@ public void retryTask(String taskGuid) throws AtlasBaseException {
...
+                task.setAttemptCount(0);
+                task.setGuid(UUID.randomUUID().toString());
+                task.setCreatedBy(RequestContext.getCurrentUser());
+
+                createTaskVertex(task);
+                createdTasks.add(task);
+            }
+        }
+
+        graph.commit();
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/tasks/AtlasTaskService.java:186]: Do we need graph.commit();, given we are using graphTransaction annotation

File: repository/src/main/java/org/apache/atlas/tasks/AtlasTaskService.java [hunk 4]
```diff
@@ -124,6 +133,197 @@ public void retryTask(String taskGuid) throws AtlasBaseException {
...
+                    deleteTask(taskGuid);
+
+                    deletedTasks.add(task);
+                } catch (AtlasBaseException e) {
+                    LOG.error("Failed to delete task {}", taskGuid);
+                }
+            }
+        }
+
+        graph.commit();
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/tasks/AtlasTaskService.java:228]: Shoudn't we use graphTransaction annotation?

File: repository/src/main/java/org/apache/atlas/tasks/AtlasTaskService.java (hunk 1 above)
Comment (sumandas0) [reply repository/src/main/java/org/apache/atlas/tasks/AtlasTaskService.java]: Done

File: repository/src/main/java/org/apache/atlas/tasks/AtlasTaskService.java (hunk 2 above)
Comment (sumandas0) [reply repository/src/main/java/org/apache/atlas/tasks/AtlasTaskService.java]: Added @Graphtransaction annotation

File: repository/src/main/java/org/apache/atlas/tasks/AtlasTaskService.java (hunk 3 above)
Comment (sumandas0) [reply repository/src/main/java/org/apache/atlas/tasks/AtlasTaskService.java:186]: Addressed

File: repository/src/main/java/org/apache/atlas/tasks/AtlasTaskService.java (hunk 4 above)
Comment (sumandas0) [reply repository/src/main/java/org/apache/atlas/tasks/AtlasTaskService.java]: Addressed

File: repository/src/main/java/org/apache/atlas/tasks/AtlasTaskService.java (hunk 1 above)
Comment (sumandas0) [reply repository/src/main/java/org/apache/atlas/tasks/AtlasTaskService.java]: Done

File: addons/policies/bootstrap_admin_policies.json [hunk 5]
```diff
@@ -24,6 +24,31 @@
           "admin-repair-index"
         ]
       }
+    },
+    {
+      "typeName": "AuthPolicy",
+      "attributes": {
+        "name": "DENY_TASK_CUD_DENY",
```
Comment (mehtaanshul) [inline addons/policies/bootstrap_admin_policies.json:31]: ALLOW_TASK_CUD

File: addons/policies/bootstrap_admin_policies.json [hunk 6]
```diff
@@ -24,6 +24,31 @@
...
+          "service-account-atlan-backend"
+        ],
+        "policyGroups": [],
+        "policyRoles": [],
+        "policyResourceCategory": "ADMIN",
+        "policyResources": [
+          "atlas-service:*"
+        ],
+        "policyActions": [
+          "api-task-cud"
```
Comment (mehtaanshul) [inline addons/policies/bootstrap_admin_policies.json:49]: "admin-task-cud"

File: auth-agents-common/src/main/resources/service-defs/atlas-servicedef-atlas.json [hunk 7]
```diff
@@ -115,7 +115,7 @@
 			},
 			"label": "Atlas Service",
 			"description": "Atlas Service",
-			"accessTypeRestrictions": ["admin-import", "admin-export", "admin-purge", "admin-audits", "admin-entity-audits", "admin-repair-index"]
+			"accessTypeRestrictions": ["admin-import", "admin-export", "admin-purge", "admin-audits", "admin-entity-audits", "admin-repair-index", "api-create-task"]
```
Comment (mehtaanshul) [inline auth-agents-common/src/main/resources/service-defs/atlas-servicedef-atlas.json:118]: "admin-task-cud"

File: auth-agents-common/src/main/resources/service-defs/atlas-servicedef-atlas.json [hunk 8]
```diff
@@ -445,6 +445,11 @@
 			"itemId": 23,
 			"name": "admin-repair-index",
 			"label": "Admin Repair Index"
+		},
+		{
+			"itemId": 24,
```
Comment (mehtaanshul) [inline auth-agents-common/src/main/resources/service-defs/atlas-servicedef-atlas.json:450]: "admin-task-cud"

File: authorization/src/main/java/org/apache/atlas/authorize/AtlasPrivilege.java [hunk 9]
```diff
@@ -49,8 +49,9 @@ public enum AtlasPrivilege {
      ADMIN_AUDITS("admin-audits"),
 
      ADMIN_ENTITY_AUDITS("admin-entity-audits"),
-     ADMIN_REPAIR_INDEX("admin-repair-index");
+     ADMIN_REPAIR_INDEX("admin-repair-index"),
 
+     API_CUD_TASK("api-task-cud");
```
Comment (mehtaanshul) [inline authorization/src/main/java/org/apache/atlas/authorize/AtlasPrivilege.java:54]: "admin-task-cud"

File: addons/policies/bootstrap_admin_policies.json (hunk 5 above)
Comment (sumandas0) [reply addons/policies/bootstrap_admin_policies.json]: Addressed

File: addons/policies/bootstrap_admin_policies.json (hunk 6 above)
Comment (sumandas0) [reply addons/policies/bootstrap_admin_policies.json]: Addressed

File: auth-agents-common/src/main/resources/service-defs/atlas-servicedef-atlas.json (hunk 7 above)
Comment (sumandas0) [reply auth-agents-common/src/main/resources/service-defs/atlas-servicedef-atlas.json]: Addressed

File: auth-agents-common/src/main/resources/service-defs/atlas-servicedef-atlas.json (hunk 8 above)
Comment (sumandas0) [reply auth-agents-common/src/main/resources/service-defs/atlas-servicedef-atlas.json:450]: Addressed

File: authorization/src/main/java/org/apache/atlas/authorize/AtlasPrivilege.java (hunk 9 above)
Comment (sumandas0) [reply authorization/src/main/java/org/apache/atlas/authorize/AtlasPrivilege.java]: Addressed

File: repository/src/main/java/org/apache/atlas/tasks/AtlasTaskService.java [hunk 10]
```diff
@@ -121,7 +131,190 @@ public void retryTask(String taskGuid) throws AtlasBaseException {
...
+        // Get all the supported task type dynamically
+        Reflections reflections = new Reflections("org.apache.atlas.repository.store.graph.v2.tasks", new SubTypesScanner(false));
+        Set<Class<? extends TaskFactory>> classes = reflections.getSubTypesOf(TaskFactory.class);
+        for (Class<?> clazz : classes) {
+            List<Class<?>> interfaces = Arrays.asList(clazz.getInterfaces());
+            if (interfaces.contains(TaskFactory.class)) {
+                try {
+                    Field field = clazz.getField("supportedTypes");
+                    List<String> supportedTaskTypes = (List<String>) field.get(null);
+                    supportedTypes.addAll(supportedTaskTypes);
```
Comment (nikhilbonte21) [inline repository/src/main/java/org/apache/atlas/tasks/AtlasTaskService.java:150]: Why fetching from reflections? the scope is limited as there are only a couple of implementations, can we populate this directly using lists in both implementation?

Comment (sumandas0) [reply repository/src/main/java/org/apache/atlas/tasks/AtlasTaskService.java:150]: Any Downside of this?

Comment (sumandas0) [reply repository/src/main/java/org/apache/atlas/tasks/AtlasTaskService.java:150]: Bit of performance issue but it shouldn't affect much I think

File: repository/src/main/java/org/apache/atlas/tasks/AtlasTaskService.java [hunk 11]
```diff
@@ -121,7 +131,190 @@ public void retryTask(String taskGuid) throws AtlasBaseException {
...
+    @GraphTransaction
+    public List<AtlasTask> createAtlasTasks(List<AtlasTask> tasks) throws AtlasBaseException {
+        AtlasPerfMetrics.MetricRecorder metric = RequestContext.get().startMetricRecord("createAtlasTasks");
+        List<String> supportedTypes = new ArrayList<>();
+        // Get all the supported task type dynamically
+        Reflections reflections = new Reflections("org.apache.atlas.repository.store.graph.v2.tasks", new SubTypesScanner(false));
+        Set<Class<? extends TaskFactory>> classes = reflections.getSubTypesOf(TaskFactory.class);
+        for (Class<?> clazz : classes) {
+            List<Class<?>> interfaces = Arrays.asList(clazz.getInterfaces());
+            if (interfaces.contains(TaskFactory.class)) {
```
[the remark this answers is not in the document]
Comment (sumandas0) [reply repository/src/main/java/org/apache/atlas/tasks/AtlasTaskService.java:146]: We can remove this check but added in some case some implemented it in nested fashion, thats it.  Its just a double verification as I was a bit less confident on reflections.

File: repository/src/main/java/org/apache/atlas/tasks/AtlasTaskService.java (hunk 10 above)
Comment (nikhilbonte21) [reply repository/src/main/java/org/apache/atlas/tasks/AtlasTaskService.java:150]: I suggest removing the reflection to prepare allow types list, I do not think we will use this in the near future, we can edit to add to the list when needed.
Also missing this in the future will not cause any adverse effect other than an error message

Comment (sumandas0) [reply repository/src/main/java/org/apache/atlas/tasks/AtlasTaskService.java]: Yep gotcha!

**Review by nikhilbonte21 — APPROVED** [review]

**Pull request merged**